r1cs/gadgets/std/src/groups/nonnative/nonnative_group_gadget.rs

```diff
@@ -0,0 +1,844 @@
+    }
+
+    fn cost() -> usize {
+        2 * <NonNativeFieldGadget<SimulationF, ConstraintF> as CondSelectGadget<ConstraintF>>::cost() +
+            <Boolean as CondSelectGadget<ConstraintF>>::cost()
+    }
+}
+
+
+
+impl<P, ConstraintF, SimulationF> ConstantGadget<SWProjective<P>, ConstraintF>
+for GroupAffineNonNativeGadget<P, ConstraintF, SimulationF>
+    where
+        P: SWModelParameters<BaseField = SimulationF>,
+        ConstraintF: PrimeField,
+        SimulationF: PrimeField + SquareRootField
+{
+    fn from_value<CS: ConstraintSystem<ConstraintF>>(
+        mut cs: CS,
+        value: &SWProjective<P>,
+    ) -> Self
+    {
+        let value = value.into_affine();
+        let x = NonNativeFieldGadget::from_value(cs.ns(|| "hardcode x"), &value.x);
+        let y = NonNativeFieldGadget::from_value(cs.ns(|| "hardcode y"), &value.y);
+        let infinity = Boolean::constant(value.infinity);
+
+        Self::new(x, y, infinity)
+    }
+
+    fn get_constant(&self) ->SWProjective<P> {
+        let value_proj = SWAffine::<P>::new(
+            self.x.get_value().unwrap(),
+            self.y.get_value().unwrap(),
+            self.infinity.get_value().unwrap()
+        ).into_projective();
+        let x = value_proj.x;
+        let y = value_proj.y;
+        let z = value_proj.z;
+        SWProjective::<P>::new(x, y, z)
+    }
+}
+
+impl<P, ConstraintF, SimulationF> AllocGadget<SWProjective<P>, ConstraintF>
+for GroupAffineNonNativeGadget<P, ConstraintF, SimulationF>
+    where
+        P: SWModelParameters<BaseField = SimulationF>,
+        ConstraintF: PrimeField,
+        SimulationF: PrimeField + SquareRootField
+{
+    #[inline]
+    fn alloc<FN, T, CS: ConstraintSystem<ConstraintF>>(
+        mut cs: CS,
+        value_gen: FN,
+    ) -> Result<Self, SynthesisError>
+        where
+            FN: FnOnce() -> Result<T, SynthesisError>,
+            T: Borrow<SWProjective<P>>,
+    {
+        let (x, y, infinity) = match value_gen() {
+            Ok(ge) => {
+                let ge = ge.borrow().into_affine();
+                (Ok(ge.x), Ok(ge.y), Ok(ge.infinity))
+            },
+            _ => (
+                Err(SynthesisError::AssignmentMissing),
+                Err(SynthesisError::AssignmentMissing),
+                Err(SynthesisError::AssignmentMissing),
+            ),
+        };
+
+        // Perform on-curve check.
+        let b = P::COEFF_B;
+        let a = P::COEFF_A;
+
+        let x = NonNativeFieldGadget::alloc(&mut cs.ns(|| "x"), || x)?;
+        let y = NonNativeFieldGadget::alloc(&mut cs.ns(|| "y"), || y)?;
+        let infinity = Boolean::alloc(&mut cs.ns(|| "infinity"), || infinity)?;
+
+        // Check that y^2 = x^3 + ax +b
+        // We do this by checking that y^2 - b = x * (x^2 +a)
+        // let x2 = x.square(&mut cs.ns(|| "x^2"))?;
+        // let y2 = y.square(&mut cs.ns(|| "y^2"))?;
+        let x2 = x.mul(cs.ns(|| "x^2"), &x)?;
+        let y2 = y.mul(cs.ns(|| "y^2"), &y)?;
```

Comment:
Let us check the number of constraints when not reducing y2 right after the product. 
We only do an addition by -b afterwards, before doing the conditional_enforce_equal.